exonum/src/encoding/protobuf/mod.rs

```diff
@@ -0,0 +1,245 @@
+
+    fn to_pb_field(&self) -> Self::ProtoValue {
+        self.to_pb()
+    }
+
+    fn from_pb_field(pb: Self::ProtoValue) -> Result<Self, ()> {
+        Self::from_pb(pb)
+    }
+}
+
+impl ToProtobuf for crypto::Hash {
+    type ProtoStruct = Hash;
+
+    fn to_pb(&self) -> Hash {
+        let mut hash = Hash::new();
+        hash.set_data(self.as_ref().to_vec());
+        hash
+    }
+
+    fn from_pb(pb: Hash) -> Result<Self, ()> {
+        let data = pb.get_data();
+        crypto::Hash::from_slice(data).ok_or(())
+    }
+}
+
+impl ToProtobuf for crypto::PublicKey {
+    type ProtoStruct = PublicKey;
+
+    fn to_pb(&self) -> PublicKey {
+        let mut key = PublicKey::new();
+        key.set_data(self.as_ref().to_vec());
+        key
+    }
+
+    fn from_pb(pb: PublicKey) -> Result<Self, ()> {
+        let data = pb.get_data();
+        crypto::PublicKey::from_slice(data).ok_or(())
+    }
+}
+
+impl ToProtobuf for bit_vec::BitVec {
+    type ProtoStruct = BitVec;
+
+    fn to_pb(&self) -> BitVec {
+        let mut bit_vec = BitVec::new();
+        bit_vec.set_data(self.to_bytes());
+        bit_vec.set_len(self.len() as u64);
+        bit_vec
+    }
+
+    fn from_pb(pb: BitVec) -> Result<Self, ()> {
+        let data = pb.get_data();
+        let mut bit_vec = bit_vec::BitVec::from_bytes(data);
+        bit_vec.truncate(pb.get_len() as usize);
+        Ok(bit_vec)
+    }
+}
+
+impl ToProtobuf for DateTime<Utc> {
+    type ProtoStruct = well_known_types::Timestamp;
+
+    fn to_pb(&self) -> well_known_types::Timestamp {
+        let mut ts = well_known_types::Timestamp::new();
+        ts.set_seconds(self.timestamp());
+        ts.set_nanos(self.timestamp_subsec_nanos() as i32);
+        ts
+    }
+
+    fn from_pb(pb: well_known_types::Timestamp) -> Result<Self, ()> {
+        Ok(Utc.timestamp(pb.get_seconds(), pb.get_nanos() as u32))
+    }
+}
+
+impl ProtobufValue for String {
+    type ProtoValue = Self;
+    fn to_pb_field(&self) -> Self::ProtoValue {
+        self.clone()
+    }
+    fn from_pb_field(pb: Self::ProtoValue) -> Result<Self, ()> {
+        Ok(pb)
+    }
+}
+
+impl ProtobufValue for Height {
+    type ProtoValue = u64;
+    fn to_pb_field(&self) -> Self::ProtoValue {
+        self.0
+    }
+    fn from_pb_field(pb: Self::ProtoValue) -> Result<Self, ()> {
+        Ok(Height(pb))
+    }
+}
+
+impl ProtobufValue for Round {
+    type ProtoValue = u32;
+    fn to_pb_field(&self) -> Self::ProtoValue {
+        self.0
+    }
+    fn from_pb_field(pb: Self::ProtoValue) -> Result<Self, ()> {
+        Ok(Round(pb))
+    }
+}
+
+impl ProtobufValue for ValidatorId {
+    type ProtoValue = u32;
+    fn to_pb_field(&self) -> Self::ProtoValue {
+        u32::from(self.0)
+    }
+    fn from_pb_field(pb: Self::ProtoValue) -> Result<Self, ()> {
+        if pb <= u32::from(u16::max_value()) {
+            Ok(ValidatorId(pb as u16))
+        } else {
+            Err(())
+        }
+    }
+}
+
+impl ProtobufValue for u32 {
```

Comment:
Is the idea behind such trivial implementations that we will use some way to auto-generate `ToProtobuf` implementations in the future?

Reply:
Yes. So that conversion functions looks the same.
```
fn from_pb(pb: Message) -> Result<Self, E> {
  Self {
    field: ProtobufValue: from_pb(pb.take_field())?
    ...
  }
}
```
And similar for conversions in other direction.

Reply:
To be really nitpicky, the current code uses `get_field` / `take_field` accessors, so the code isn't completely same for all fields.

Reply:
Yes, because it is impossible to take elementary types, but I think that it is going to be a simple special case. But for this to be simple one have to use specific message types for ValidatorId, Round, etc(than only simple types such as u64, etc. will use get_field)
I did not think any other way to generate simple conversions. 

Lets say that we want to use get_field method that is defined for every field type. But problem is that get_field can return different type (`&str` for `String` type, slice for vector, etc.) One can go around this but maybe I thought that it is harder to achieve than current approach.

But we can use fields that is stored in pb structs directly. But they can be stored in different wrappers. (u64 is stored like u64, but `Hash` is stored as `SingularField<Hash>`)

Reply:
I plan to provide auto-derive for trait `ToProtobuf` which will generate standard conversions. They will look something like this.
```
fn from_pb(pb: Message) -> Result<Self, E> {
  Self {
    field: ProtobufValue: from_pb(pb.get_field().to_owned())?
    ...
  }
}

fn tp_pb(&self) ->Message {
    ...
    message.set_field(self.field.to_owned())
    ...
    Ok(message)
}
```